### ISSUE TITLE
Provide a way to force set an environment variable from phpunit.xml

### ISSUE THREAD

**Issue by julianxhokaxhiu** [sebastianbergmann/phpunit#2353]
As already mentioned in https://github.com/sebastianbergmann/phpunit/issues/1276 there should be one way to override already defined ENV variabiles.

Looking at this line ( https://github.com/sebastianbergmann/phpunit/blob/5.6/src/Util/Configuration.php#L542 ) there's still a check that defines it, only if undefined. But there are some other cases where should be possible to override them.

My proposal:
```xml
<env name="FOO" value="bar" force="true" />
```

if `force` is set to `true` the env will be always set, no matter what.

Thanks in advance,
Julian

**Comment by FlashBlack** [comment]
+1

There is no need to prohibit the overriding of environment variables.

**Comment by cxj** [comment]
I just spent several hours debugging my unit tests because of this behavior.  PHPUnit will _not set_ an existing environment variable, even if it is **empty**.  That is, this will prevent that carefully crafted `phpunit.xml` clause from setting an environment variable named `APP_HOME`:

```sh
# sh syntax
export APP_HOME=

# Note:
~$ echo $APP_HOME

~$ echo $DOES_NOT_EXIST_AND_IS_NOT_SET

~$ set | egrep 'APP_HOME|DOES_NOT_EXIST'
APP_HOME=
```

Or from PHP (file bar.php):
```php
<?php
echo 'APP_HOME = ' . getenv('APP_HOME') . PHP_EOL;
echo 'DOES_NOT_EXIST_AND_IS_NOT_SET = ' . getenv('DOES_NOT_EXIST_AND_IS_NOT_SET') . PHP_EOL;
?>
```

Output:
```sh
~$ php bar.php
APP_HOME =
DOES_NOT_EXIST_AND_IS_NOT_SET =
```

Thus, gyrations are needed within the test classes themselves:
```php
        if (strlen(getenv('APP_HOME')) < 1) {
            echo PHP_EOL . 'Error:  APP_HOME set but empty' . PHP_EOL;
            $this->stopFlag = true;
        }
```
```php
    public function run(\PHPUnit_Framework_TestResult $result = null)
    {
        $result = parent::run($result);

        if ($this->stopFlag === true)
        {
            exit(1);    // Actually does exit here.
        }
        return $result; // Return as normal.
    }
```

My tests once again work after this.  But it would be nice to have a feature to allow PHPUnit to forcibly set required environment variables to known, sane starting values!

**Comment by andrewmclagan** [comment]
I agree this seems out-of-line with expected behaviour. Otherwise why would we not just set environment variables?

**Comment by mcustiel** [comment]
A comment in the code is pointing out that this "not overwrite feature" was added because of this: https://github.com/sebastianbergmann/phpunit/issues/1181
Would it make sense to have an overwrittable flag in the config? Because I see the point on each use case

**Comment by mcustiel** [comment]
After discussing the issue with @sebastianbergmann, it was decided that it does make sense to have this check to allow what it was requested in https://github.com/sebastianbergmann/phpunit/issues/1181
A user should be allowed to set an environment variable in the command line as it is usually being done when running CLI applications (see the example in #1181). So the idea is to have the configuration file to set the global environment to run the tests, but they don't have priority over variables that come from outside the application so we can be able to set variables on the fly for the CLI to run.
If it is really important to overwrite a var (let's say $HOME), maybe it could be done in the bootstrap file?

**Comment by helhum** [comment]
After giving this some thought as well, I would agree with @mcustiel that overriding env vars in the config file shouldn't be necessary, or if so, can be done in the configured bootstrap.

@sebastianbergmann close this one then?

**Comment by Jonnx** [comment]
@sebastianbergmann  @helhum i would disagree with the attempt to close this. allowing the override of environment variables would massively simplify running tests and developing locally with docker and using the same local configuration for tests during CI / CD

**Comment by mfn** [comment]
I was bitten by this too, after upgrading to phpunit 6.2.1 from some 5.7.20.

I now had to prepend my shell command with `APP_ENV=testing .../phpunit` were previously I just were able to call `.../phpunit`. Turned out the reason was that our CI provide Codeship defines `APP_ENV=CODESHIP` but our test suite has a common dependency on `testing` and doesn't need special treatment for the CI environment.


TL;DR: I would have expected `<env ...>` defined in `phpunit.xml` to be authoritative and override any shell environments being inherited to phpunit.

**Comment by tdmalone** [comment]
Another +1 from me here.

Ran into this today and fortunately it didn't take too long to work out what was happening.

Even though I already have `APP_ENV` set to `'testing'` in my `phpunit.xml.dist`, my local test command has now gone from this

    /vendor/bin/phpunit

to this

    BACKUP_APP_ENV=$APP_ENV; APP_ENV=testing; vendor/bin/phpunit; APP_ENV=$BACKUP_APP_ENV

**Comment by julianxhokaxhiu** [comment]
I really don't get, why is it so difficult to introduce one flag, that is up to the developer to decide if to use it or not?

**Comment by sebastianbergmann** [comment]
@julianxhokaxhiu The solution proposed in #2401 was too complicated. Submit a simpler pull request and I'll have a look.

**Comment by julianxhokaxhiu** [comment]
I'll do, thanks

**Comment by julianxhokaxhiu** [comment]
Done. Unit Tests passes as well :)
